server/src/core/payment/transaction-service.ts

```diff
@@ -231,6 +217,52 @@ export class Transactions implements TransactionService {
     }
   }
 
+  private async getConfirmedUserBalance(userId: string): Promise<number> {
+    try {
+      const results = await this.collection.aggregate<{ balance: number }>([
+        {
+          // get transactions to or from this user
+          $match: {
+            $or: [
+              // be conservative when dealing with pending transactions:
+              // if transaction is from the user, deduct it from balance unless it's failed
+              // if transaction is to user, add it to balance only if it's successful
+              { from: userId, status: { $ne: 'failed' } },
```

Comment:
@BambanzaJuniorThe2nd  in that case, the system won't retry the transaction, but the status of the transaction will be set to `failed`. And the next time the balance is computed, that transaction won't be taken into account since it would now be `failed`.